[Feature] 관리자 회원가입, 로그인 기능을 구현한다.

ghrltjdtprbs opened the pull request:
## ⭐️ [Feature] 관리자 회원가입, 로그인 기능을 구현한다.

- 관리자 회원가입을 구현하였습니다.
- 관리자 로그인을 구현하였습니다.
- REST Docs 문서화를 구현하였습니다. 


## 관련 이슈

- #81 

## 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [ ] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트

## 작업 기간

- 작업 시작일: (2024-05-20)
- 작업 종료일: (2024-05-21)

ghrltjdtprbs committed:
[feat] jackson-datatype-jsr310 모듈 등록

변경 사항 :
 - LocalDateTime의 직렬화, 역직렬화를 위해 jackson-datatype-jsr310 의존성 추가 및 수동 구성

ghrltjdtprbs committed:
[style] 이메일 중복 예외 클래스 이름 변경

문제 사항 :
 - 이메일 중복 예외가 여러군데 쓰일 수 있음
 - MemberDuplicatedException 클래스명은 부적절하다고 판단.

변경 사항 :
 - MemberDuplicatedException -> EmailDuplicatedException

ghrltjdtprbs committed:
[feat] OfficeNotFoundException 예외 추가

변경 사항 :
 - ErrorCode에 OFFICE_NOT_FOUND 추가
 - OfficeNotFoundException 클래스 추가

ghrltjdtprbs committed:
[feat] 관리자 회원가입, 로그인 추가

변경 사항 :
 - 관리자 회원가입, 로그인 기능을 추가하였습니다.

ghrltjdtprbs committed:
[style] 오타 수정

변경 사항 :
 - 칼럼 오타 수정 adress -> address

ghrltjdtprbs committed:
[refactor] 클래스명 변경 반영

문제 사항 :
 - EmailDuplicatedException로 클래스 명을 변경하여 반영 필요

변경 사항 :
 - import문, 예외 처리 변경

ghrltjdtprbs committed:
[feat] requestMatchers 추가

변경 사항 :
 - "/members/**" 삭제
 - "/members/login" 추가

ghrltjdtprbs committed:
[feat] 관리자 REST Docs문서화

변경 사항 :
 - AdminControllerDocsTest 추가
 - .adoc파일 추가

ghrltjdtprbs committed:
[refactor] 통일성을 위한 member-api.adoc 수정

변경 사항 :
 - 대소문자 구분
 - response-headers.adoc 추가

Galmaeki commented on src/main/java/com/final_10aeat/global/config/SecurityConfig.java:
```diff
@@ -34,9 +34,9 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
-                    .requestMatchers(HttpMethod.GET,"/health/**").permitAll()
-                    .requestMatchers("/members/**").permitAll()//임시허용
+                    .requestMatchers(HttpMethod.POST, "/members/register", "/members/login",
+                        "/admin").permitAll()
```
뒤에 와일드카드 없어도 허용 되나요?

ghrltjdtprbs replied:
관리자 관련 api 경로에 전부 /admin이 붙어서 일부러 "/admin"로 명시하였습니다. 😊

Galmaeki commented on src/main/java/com/final_10aeat/domain/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,39 @@
+    private final AdminService adminService;
+
+    @PostMapping
+    public ResponseEntity<ResponseDTO<Void>> registerAdmin(
```
컨트롤러에서 Admin이 표기되어있으니 register만 사용해도 좋다고 생각합니다

ghrltjdtprbs replied:
생각해보니 그게 더 깔끔할 것 같네요! 수정해서 커밋했으니 확인해주세요 ㅎㅎ
관련 커밋 : 43ab74f0629e505b2781b35717ba6445ba926bce

Galmaeki commented on src/main/java/com/final_10aeat/domain/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,39 @@
+    @PostMapping("/login")
+    public ResponseDTO<Void> login(
+        HttpServletResponse response,
+        @RequestBody @Valid MemberLoginRequestDto request) {
```
MemberLoginRequestDto를 common이나 global로 빼도 좋을것 같습니다!

ghrltjdtprbs replied:
제 개인적인 생각으로는 로그인에서 두 번 사용이 끝이고, 앞으로 사용 가능성이 없기 때문에 따로 빼기에는 부적절한 것 같습니다!

Galmaeki commented on src/main/java/com/final_10aeat/domain/admin/dto/request/CreateAdminRequestDto.java:
```diff
@@ -0,0 +1,29 @@
+import java.time.LocalDateTime;
+
+public record CreateAdminRequestDto(
+    @Email(message = "이메일 형식이 유효하지 않습니다.")
```
Email 어노테이션은 null이 허용됩니다!
DB단에서도 막히긴 하지만 NotBlank도 같이 사용해주셨으면 좋을 것 같습니다

ghrltjdtprbs replied:
모르고 있었네요... 좋은 정보 감사합니다! 수정했으니 확인해주세요!
관련 커밋 : 76c9e8a950209b02b0d0d46745d682a7a221086e

Galmaeki commented on src/main/java/com/final_10aeat/domain/admin/service/AdminService.java:
```diff
@@ -0,0 +1,74 @@
+    private final PasswordEncoder passwordEncoder;
+    private final JwtTokenGenerator jwtTokenGenerator;
+
+    public Admin registerAdmin(CreateAdminRequestDto request) {
```
이부분도 Admin이 Class이름에 존재하니 Admin을 제거해도 좋을것같습니다

ghrltjdtprbs replied:
수정해서 커밋했습니다! 😊
관련 커밋 : 43ab74f0629e505b2781b35717ba6445ba926bce

Galmaeki commented on src/main/java/com/final_10aeat/domain/admin/service/AdminService.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private void validateEmailNotDuplicated(String email) {
+        if (adminRepository.findByEmail(email).isPresent()) {
```
existByEmail을 사용하지 않은 이유가 있을까요?

ghrltjdtprbs replied:
Optional을 사용하여 이메일의 존재 여부를 확인하고 예외를 던지는 방식은 명확하고 일관된 예외 처리를 가능하게 합니다! 
그리고 현재 로직에서는 findByEmail을 사용해도 성능에 큰 차이가 없고, 후에 admin로직의 findByEmail을 사용할 일이 생길 것 같아 existByEmail 대신 findByEmail를 사용하였습니다! 👍👍

ghrltjdtprbs committed:
[Rename] 경로 수정

변경 사항 :
 - AdminControllerDocsTest 경로 수정

Galmaeki commented on src/test/java/com/final_10aeat/domain/admin/AdminControllerDocsTest.java:
```diff
@@ -0,0 +1,162 @@
+        );
+
+        // When
+        when(adminService.registerAdmin(registerRequest)).thenReturn(null);
```
이 부분은 메소드 이름은 when이지만 given 부분입니다!
아래 주석을 // When & Then 으로 사용하는게 더 BDD 패턴에 맞는 사용법입니다

ghrltjdtprbs replied:
수정해서 커밋했습니다! 지적 감사합니다 ㅎㅎ😊
관련 커밋 : f053bdbd80c0f8ad13c1bb2a48d6ef5aed774245

Galmaeki reviewed (changes requested):
고생 많으셨습니다!
코멘트 확인 부탁드리겠슴다!

ghrltjdtprbs committed:
[feat] AdminServiceTest 추가

변경 사항 :
 - AdminServiceTest 추가

ghrltjdtprbs committed:
[refactor] MemberServiceTest 메서드명 변경

변경 사항 :
 - code convention에 맞게 메서드명 변경

ghrltjdtprbs committed:
[test] AdminEntityTest 테스트 추가

ghrltjdtprbs committed:
[feat] 테스트 환경에서 H2 데이터베이스 사용 설정

문제 사항 :
 - 테스트 코드가 프로젝트의 실제 데이터베이스 설정과 독립적으로 동작할 수 있게 되어, 더 안정적이고 신뢰성 있는 테스트 환경을 구축할 필요가 있음.

변경 사항 :
 - build.gradle에 'testImplementation 'com.h2database:h2'' 의존성 추가
 - 테스트 코드에서 H2 데이터베이스를 사용하여 독립적인 테스트 환경 구성
 - 프로젝트의 실제 데이터베이스는 MySQL을 사용하며, 테스트 환경에서는 H2를 사용하여 빠르고 간편한 테스트 수행 가능

ghrltjdtprbs committed:
[refactor] CreateAdminRequestDto 이메일 검증추가

문제 사항 :
 - @Email 어노테이션은 null이 허용되는 문제

변경 사항 :
 - @NotBlank 를 추가해 null 방지

ghrltjdtprbs committed:
[refactor] registerAdmin 메소드명 변경

문제 사항 :
 - registerAdmin 보다 register가 더 적절하다는 코드리뷰 반영

변경 사항 :
 - registerAdmin -> register 메소드명 변경
 - 관련 코드 모두 수정

ghrltjdtprbs committed:
[feat] requestMatchers에 "/admin/login"추가

문제 사항 :
 - 관리자 로그인 또한 로그인 하지 않은 상태에서 수행해야함

변경 사항 :
 - requestMatchers에 "/admin/login"추가

ghrltjdtprbs committed:
[refactor] testRegisterAdmin 주석 수정

문제 사항 :
 - // When 주석이 달린 부분이 사실 given 부분에 포함되어 주석이 올바르지 못함

변경 사항 :
 - 주석 수정

asd42270 reviewed (approved):
고생하셨습니다!!! 재현님이 다 말씀해 주신 것 같아서 저는 딱히 없는 것 같아욤

Galmaeki reviewed (approved):
고생 많으셨슴다!
